### PR TITLE
[RFC] Decoupling: microVM lifetime clearly split between pre-boot and post-boot.

### PR DESCRIPTION
**Pull request by acatangiu**
## Description of Changes

Disclaimer: This is a rather heavy PR/RFC, but benefits should be obvious.
Note: I've kept the gritty commit history in the hope that it would be easier to review, I suggest going through _all_ commits descriptions so you get an idea of where things are going, then you can review the final version or all the twists and turns the code took to get there.

The microVM lifetime and how resources are owned and managed is now clearly split in two:
1. There is a pre-boot *vmm-building* loop where pre-boot-supported API requests are handled. These mostly work on the resource collection defining the microVM. At this time, Firecracker acts as a data-store exposing a REST API for managing the data (microVM resources). A successful `StartMicroVm` command (successful boot) will move the microVM into the post-boot state, while failures are recoverable as they do not dirty/mutate the data.
2. There is a post-boot `VmController` which runs a control loop that handles `Vmm` internal events and also handles post-boot-supported API requests.

There is a very clear separation of concerns between objects used in each of these consecutive stages.

Another clear delimitation which occurred is between:
1. Configuring from API
and
2. Configuring from cmdline JSON

Another one is between a microVM with its API enabled or disabled.

The `VmController` was further enhanced to provide a generic interface taking an external `trait` object for externally handling control events.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/vmm/src/controller.rs [hunk 1]
```diff
@@ -57,19 +57,6 @@ impl VmmController {
             .started
     }
 
-    /// Sets a vsock device to be attached when the VM starts.
-    pub fn set_vsock_device(&mut self, config: VsockDeviceConfig) -> UserResult {
-        if self.is_instance_initialized() {
```
Comment (raduiliescu) [inline src/vmm/src/controller.rs]: I don't think it is OK to delete this check. This function is also called as an external API call and can be called anytime.

File: src/vmm/src/builder.rs [hunk 2]
```diff
@@ -169,6 +170,29 @@ impl VmmBuilder {
             .map_err(VmmActionError::from)
     }
 
+    /// Updates the path of the host file backing the emulated block device with id `drive_id`.
+    pub fn update_block_device_path(&mut self, drive_id: String, path_on_host: String) -> UserResult {
```
Comment (raduiliescu) [inline src/vmm/src/builder.rs]: 2 public functions with same name... also not clear what you do with this function. Isn't this a leftover?

Comment (raduiliescu) [reply src/vmm/src/builder.rs]: I figured it out what you have done here and is fine.

File: src/vmm/src/controller.rs (hunk 1 above)
Comment (raduiliescu) [reply src/vmm/src/controller.rs]: Please note you haven't made vsock config as OperationNotSupportedPostBoot.

Comment (acatangiu) [reply src/vmm/src/controller.rs]: Good catch!

Comment (acatangiu) [reply src/vmm/src/controller.rs]: Actually, I've checked and all external API calls are covered in both pre-boot and post-boot handlers. They actually **have to** be covered as a `match` in rust doesn't allow incomplete coverage of a matched value's range.

**Pull request closed without merging**